lighthouse-core/audits/seo/tap-targets.js

```diff
@@ -0,0 +1,274 @@
+    }
+  }
+  return true;
+}
+
+/**
+ *
+ * @param {LH.Artifacts.TapTarget[]} targets
+ */
+function getTooSmallTargets(targets) {
+  return targets.filter(targetIsTooSmall);
+}
+
+/**
+ *
+ * @param {LH.Artifacts.TapTarget[]} tooSmallTargets
+ * @param {LH.Artifacts.TapTarget[]} allTargets
+ */
+function getOverlapFailures(tooSmallTargets, allTargets) {
+  /** @type {LH.Audit.TapTargetOverlapDetail[]} */
+  let failures = [];
+
+  tooSmallTargets.forEach(target => {
+    const overlappingTargets = getTooCloseTargets(
+      target,
+      allTargets
+    );
+
+    if (overlappingTargets.length > 0) {
+      failures = failures.concat(overlappingTargets);
+    }
+  });
+
+  return failures;
+}
+
+/**
+ *
+ * @param {LH.Artifacts.TapTarget} tapTarget
+ * @param {LH.Artifacts.TapTarget[]} allTapTargets
+ */
+function getTooCloseTargets(tapTarget, allTapTargets) {
+  /** @type LH.Audit.TapTargetOverlapDetail[] */
+  const failures = [];
+
+  for (const maybeOverlappingTarget of allTapTargets) {
+    if (maybeOverlappingTarget === tapTarget) {
+      // checking the same target with itself, skip
+      continue;
+    }
+
+    const failure = getTargetTooCloseFailure(tapTarget, maybeOverlappingTarget);
+    if (failure) {
+      failures.push(failure);
+    }
+  }
+
+  return failures;
+}
+
+/**
+ * @param {LH.Artifacts.TapTarget} tapTarget
+ * @param {LH.Artifacts.TapTarget} maybeOverlappingTarget
+ * @returns {LH.Audit.TapTargetOverlapDetail | null}
+ */
+function getTargetTooCloseFailure(tapTarget, maybeOverlappingTarget) {
+  // convert client rects to unique tappable areas from a user's perspective
+  const tappableRects = getTappableRectsFromClientRects(tapTarget.clientRects);
+  const isHttpOrHttpsLink = /https?:\/\//.test(tapTarget.href);
+  if (isHttpOrHttpsLink && tapTarget.href === maybeOverlappingTarget.href) {
+    // no overlap because same target action
+    return null;
+  }
+
+  /** @type LH.Audit.TapTargetOverlapDetail | null */
+  let greatestFailure = null;
+  for (const targetCR of tappableRects) {
+    if (allRectsContainedWithinEachOther(tappableRects, maybeOverlappingTarget.clientRects)) {
+      // If one tap target is fully contained within the other that's
+      // probably intentional (e.g. an item with a delete button inside)
+      continue;
+    }
+    for (const maybeOverlappingCR of maybeOverlappingTarget.clientRects) {
+      const failure = getOverlapFailure(targetCR, maybeOverlappingCR);
+      if (failure) {
+        // only update our state if this was the biggest failure we've seen for this pair
+        if (!greatestFailure ||
+          failure.overlapScoreRatio > greatestFailure.overlapScoreRatio) {
+          greatestFailure = {
+            ...failure,
+            tapTarget,
+            overlappingTarget: maybeOverlappingTarget,
+          };
+        }
+      }
+    }
+  }
+  return greatestFailure;
+}
+
+/**
+ * @param {LH.Artifacts.Rect} targetCR
+ * @param {LH.Artifacts.Rect} maybeOverlappingCR
```

Comment:
adding explicit return here like you did above would be 👍 really helps me understand that it might not return a failure

```js
 /*
  * @returns {LH.Audit.TapTargetOverlapDetail | null}
  */
```